### PR TITLE
Community ID or a network flow hash implementation.

### LINKED ISSUE #34062: [pkg/ottl] Support community ID network flow
### Component(s)

pkg/ottl

### Is your feature request related to a problem? Please describe.

_**What is a community ID and why do we need it?**_
- It is a single unique ID based on the network flow info. It is an additional flow identifier and doesn't replace existing flow identification mechanisms already supported by the monitors. See the [specification](https://github.com/corelight/community-id-spec).
- When monitoring/analyzing network flow, for example threat hunting security use cases, it's often required to make "joins" on network source and destination info where `community_id` simplifies, also gives a better user experience when analyzing the data (aggregate by community ID to collect statistics, etc.)
- visual example
  

The feature is widely used and here some reference applications:
- Suricate, Zeek, MISP, etc..: https://zeek.org/2019/07/31/an-update-on-community-id/
- Wireshark: https://corelight.com/blog/community-id-support-for-wireshark 
- Tenzip, a security data lake: https://github.com/tenzir/tenzir/pull/525
- Elastic beats: https://www.elastic.co/guide/en/beats/filebeat/current/community-id.html
- Elasticsearch community_id: https://www.elastic.co/guide/en/beats/filebeat/current/community-id.html
- List of vendor implementations: https://github.com/corelight/community-id-spec?tab=readme-ov-file#production-implementations

### Describe the solution you'd like

Introduce a converter which calculates the community ID based on the [specification](https://github.com/corelight/community-id-spec).

### Describe alternatives you've considered

This requires a discussion of either
- implement from scratch like [Elastic beats has `communityid`](https://github.com/elastic/beats/blob/main/libbeat/common/flowhash/communityid.go)
- or use existing _active_ go library, such as https://pkg.go.dev/github.com/satta/gommunityid

### Additional context

_No response_

### Example
- Config
```

processors:
  batch:
  transform:
    log_statements:
      - context: log
        statements:
          - set(attributes["community_id"], CommunityID(attributes["source.ip"], attributes["source.port"], attributes["destination.ip"], attributes["destination.port"]))
```

- Expected output
```
Timestamp: 2018-12-13 14:51:00.3 +0000 UTC
SeverityText: Information
SeverityNumber: Info2(10)
Body: Str(Example log record)
Attributes:
     -> timestamp: Str(2025-05-21T09:25:18Z)
     -> source.ip: Str(123.124.125.126)
     -> source.port: Int(12345)
     -> destination.ip: Str(55.56.57.58)
     -> destination.port: Int(80)
     -> boolean.attribute: Bool(true)
     -> int.attribute: Int(10)
     -> double.attribute: Double(637.704)
     -> array.attribute: Slice(["many","values"])
     -> map.attribute: Map({"some.map.key":"some value"})
     -> community_id: Str(1:3vT0XsYnQuR8XHIYy6Acr7kpWGc=)
Trace ID: 5b8efff798038103d269b633813fc60c
Span ID: eee19b7ec3c1b174

```

### PR DESCRIPTION
**Pull request by mashhurs**
#### Description

Replaces https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40235

It accepts source and destination {IP, port} pair and optionally protocol (eg. TCP) and seed (eg. 1,2,etc...) parameters.
At the result it computes the has values and generates network flow (smaller to greater) ID a.k.a community ID.

#### Link to tracking issue
- Fixes #34062

#### Testing
- How do I validate if current implementation is correctly generating community ID?
Use [Elasticsearch community ID processor](https://www.elastic.co/docs/reference/enrich-processor/community-id-processor) in Kibana Dev tools, example:
```
POST /_ingest/pipeline/_simulate
{
  "pipeline": {
    "description": "Generates Community ID for network flow data",
    "processors": [
      {
        "community_id": {}
      }
    ]
  },
  "docs": [
    {
      "_index": "index",
      "_id": "id",
      "_source": {
        "source": {
          "ip": "123.124.125.126",
          "port": 12345
        },
        "destination": {
          "ip": "55.56.57.58",
          "port": 80
        },
        "network": {
          "transport": "TCP"
        }
      }
    }
  ]
}
```
image

- unit tests
```
cd /path-to/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs && go test -v -run TestCommunityID
```
- E2E test

##### Local run

- config
```
extensions:
  zpages:
    endpoint: 127.0.0.1:55679

receivers:
  otlp:
    protocols:
      http:
        endpoint: 127.0.0.1:8080

processors:
  batch:
  transform:
    metric_statements:
      - context: resource
        statements:
          - set(attributes["community_id"], CommunityID(attributes["source.ip"], attributes["source.port"], attributes["destination.ip"], attributes["destination.port"], "TCP", 2))

exporters:
  debug:
    verbosity: detailed
  elasticsearch:
    cloudid: "cloud-id"
    user: "elastic"
    password: "pwd"

service:
  telemetry:
    logs:
      level: debug
  pipelines:
    metrics:
      receivers: [otlp]
      processors: [batch, transform]
      exporters: [elasticsearch, debug]

  extensions: [zpages]
```

- Payload
```
curl --location '127.0.0.1:8080/v1/metrics' \
--header 'Content-Type: application/json' \
--data '{
  "resourceMetrics": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "my.service"
            }
          },
          {
            "key": "timestamp",
            "value": {
              "stringValue": "2018-12-01T16:17:18Z"
            }
          },
          {
            "key": "source.ip",
            "value": { "stringValue": "123.124.125.126" }
          },
          {
            "key": "source.port",
            "value": { "intValue": "12345" }
          },
          {
            "key": "destination.ip",
            "value": { "stringValue": "55.56.57.58" }
          },
          {
            "key": "destination.port",
            "value": { "intValue": "80" }
          }
        ]
      },
      "scopeMetrics": [
        {
          "scope": {
            "name": "my.library",
            "version": "1.0.0",
            "attributes": [
              {
                "key": "my.scope.attribute",
                "value": {
                  "stringValue": "some scope attribute"
                }
              }
            ]
          },
          "metrics": [
            {
              "name": "my.counter",
              "unit": "1",
              "description": "I am a Counter",
              "sum": {
                "aggregationTemporality": 1,
                "isMonotonic": true,
                "dataPoints": [
                  {
                    "asDouble": 5,
                    "startTimeUnixNano": "1544712660300000000",
                    "timeUnixNano": "1544712660300000000",
                    "attributes": [
                      {
                        "key": "my.counter.attr",
                        "value": {
                          "stringValue": "some value"
                        }
                      }
                    ]
                  },
                  {
                    "asDouble": 2,
                    "startTimeUnixNano": "1544712660300000000",
                    "timeUnixNano": "1544712660300000000",
                    "attributes": [
                      {
                        "key": "another.counter.attr",
                        "value": {
                          "stringValue": "another value"
                        }
                      }
                    ]
                  }
                ]
              }
            },
            {
              "name": "my.gauge",
              "unit": "1",
              "description": "I am a Gauge",
              "gauge": {
                "dataPoints": [
                  {
                    "asDouble": 10,
                    "timeUnixNano": "1544712660300000000",
                    "attributes": [
                      {
                        "key": "my.gauge.attr",
                        "value": {
                          "stringValue": "some value"
                        }
                      }
                    ]
                  }
                ]
              }
            },
            {
              "name": "my.histogram",
              "unit": "1",
              "description": "I am a Histogram",
              "histogram": {
                "aggregationTemporality": 1,
                "dataPoints": [
                  {
                    "startTimeUnixNano": "1544712660300000000",
                    "timeUnixNano": "1544712660300000000",
                    "count": "2",
                    "sum": 2,
                    "bucketCounts": [
                      "1",
                      "1"
                    ],
                    "explicitBounds": [
                      1
                    ],
                    "min": 0,
                    "max": 2,
                    "attributes": [
                      {
                        "key": "my.histogram.attr",
                        "value": {
                          "stringValue": "some value"
                        }
                      }
                    ]
                  }
                ]
              }
            }
          ]
        }
      ]
    }
  ]
}'
```

- Logs
```
2025-10-09T11:37:49.524-0700    debug   ottl@v0.137.0/parser.go:37      TransformContext after statement execution      {"resource": {"service.instance.id": "5224a627-dcc7-4f0a-afa0-efad6130016a", "service.name": "otelcontribcol", "service.version": "0.137.0-dev"}, "otelcol.component.id": "transform", "otelcol.component.kind": "processor", "otelcol.pipeline.id": "metrics", "otelcol.signal": "metrics", "statement": "set(resource.attributes[\"community_id\"], CommunityID(resource.attributes[\"source.ip\"], resource.attributes[\"source.port\"], resource.attributes[\"destination.ip\"], resource.attributes[\"destination.port\"], \"TCP\", 2))", "condition matched": true, "TransformContext": {"resource": {"attributes": {"service.name": "my.service", "timestamp": "2018-12-01T16:17:18Z", "source.ip": "123.124.125.126", "source.port": 12345, "destination.ip": "55.56.57.58", "destination.port": 80, "community_id": "1:7oTw+VtCU5XAuMK7fTNdGllNeUs="}, "dropped_attribute_count": 0}, "cache": {}}}
2025-10-09T11:37:49.524-0700    info    Metrics {"resource": {"service.instance.id": "5224a627-dcc7-4f0a-afa0-efad6130016a", "service.name": "otelcontribcol", "service.version": "0.137.0-dev"}, "otelcol.component.id": "debug", "otelcol.component.kind": "exporter", "otelcol.signal": "metrics", "resource metrics": 1, "metrics": 3, "data points": 4}
2025-10-09T11:37:49.524-0700    info    ResourceMetrics #0
Resource SchemaURL: 
Resource attributes:
     -> service.name: Str(my.service)
     -> timestamp: Str(2018-12-01T16:17:18Z)
     -> source.ip: Str(123.124.125.126)
     -> source.port: Int(12345)
     -> destination.ip: Str(55.56.57.58)
     -> destination.port: Int(80)
     -> community_id: Str(1:7oTw+VtCU5XAuMK7fTNdGllNeUs=)
```

#### Documentation
- changelog

### CODE REVIEW

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+			return -1
+		}
+		if a[i] > b[i] {
+			return 1
+		}
+	}
+	return 0
+}
+
+func flow(srcIPBytes net.IP, srcPortForHash uint16, dstIPBytes net.IP, dstPortForHash uint16, protolValue uint8, seedValue uint16) string {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]:
```suggestion
func flow(srcIPBytes net.IP, srcPortForHash uint16, dstIPBytes net.IP, dstPortForHash uint16, protoValue uint8, seedValue uint16) string {
```

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+			srcPortForHash, dstPortForHash = dstPortForHash, srcPortForHash
+		}
+
+		// Build the flow, format: <seed:2><protocol:1><src_ip><dst_ip><src_port:2><dst_port:2>
+		return flow(srcIPBytes, srcPortForHash, dstIPBytes, dstPortForHash, protocolValue, seedValue), nil
+	}
+}
+
+func resolveProtocolValue(protoStr string) (uint8, error) {
+	protoMap := map[string]uint8{"ICMP": 1, "TCP": 6, "UDP": 17, "RSVP": 46, "ICMP6": 58, "SCTP": 132}
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: we can make this a variable, and also use it to lookup the value at  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/43398/files#diff-e900503d7bd07a20328514b5a6da07acaaa125e664a362d8d691b16dfc667f16R78

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+			dstIPBytes = dstIPAddr.To16()
+		}
+
+		// Determine a flow direction for normalization
+		// If source IP is greater than destination IP, or if IPs are equal and source port is greater than destination port,
+		// then swap source and destination to normalize the flow direction
+		shouldSwap := false
+		if len(srcIPBytes) != len(dstIPBytes) {
+			shouldSwap = len(srcIPBytes) > len(dstIPBytes)
+		} else if cmp := compareBytes(srcIPBytes, dstIPBytes); cmp > 0 {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: Do we've any specific reason to not use the standard `bytes.Compare`?

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		}
+
+		// Get seed value (default: 0) if applied
+		seedValue := uint16(0)
+		if !seed.IsEmpty() {
+			seedInt, err := seed.Get().Get(ctx, tCtx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get seed: %w", err)
+			}
+			seedValue = uint16(seedInt)
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: I think this number conversion might overflow. The same for line 115.

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+func communityID[K any](
+	sourceIP ottl.StringGetter[K],
+	sourcePort ottl.IntGetter[K],
+	destinationIP ottl.StringGetter[K],
+	destinationPort ottl.IntGetter[K],
+	protocol ottl.Optional[ottl.StringGetter[K]],
+	seed ottl.Optional[ottl.IntGetter[K]],
+) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		srcIPValue, err := sourceIP.Get(ctx, tCtx)
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: Can we make this function a bit smaller? Maybe by extracting this long get parameters/parse/validate logic into a separate function?

**Review by edmocosta — COMMENTED** [review]
Thank you @mashhurs and I'm sorry for the late review.
I've left a few suggestions, we also need to update the [docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md) to include this new function.

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 5 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Yeah validation and param extraction can be separated, pure algo sequence will stay in the function like: `validateAndExtractParams` -> `determineFlowDirection` -> `createFlow`.
I have made it shorter.

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 4 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: I have added a validation rule to make sure we get desired byte-size/value. Do you have a better idea?

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 3 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Nope, it is because I didn't know there is `bytes.Compare` 😃

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 2 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Yup, updated!

**Comment by mashhurs** [comment]
> Thank you @mashhurs and I'm sorry for the late review. I've left a few suggestions, we also need to update the [docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md) to include this new function.

Sorry, completely forgot the doc, I have added.
Thank you for catching this up.

**Comment by mashhurs** [comment]
CIs are really funny!
If I run `make generate` and push the changes, linter complains. If I fix linter and push the change it again complains that I should run `make generate`.

Linter and and `make generate` are fighting for this space 😄 
image

Any solution for this @edmocosta ?

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	if !protocol.IsEmpty() {
+		protocolStr, err := protocol.Get().Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get protocol: %w", err)
+		}
+
+		protocolValue = protocolMap[protocolStr]
+		if protocolValue == 0 {
+			return nil, fmt.Errorf("unsupported protocol: %s", protocolStr)
+		}
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]:
```suggestion
		var ok bool 
		protocolValue, ok = protocolMap[protocolStr]
		if !ok {
			return nil, fmt.Errorf("unsupported protocol: %s", protocolStr)
		}
```

Let's check whether the element exist in the map, and not use the zero value for that.

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	if srcPort < 1 || srcPort > 65535 {
+		return nil, fmt.Errorf("source port must be between 1 and 65535, got %d", srcPort)
+	}
+
+	dstPort, err := destinationPort.Get(ctx, tCtx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get destination port: %w", err)
+	}
+	if dstPort < 1 || dstPort > 65535 {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: Shouldn't these port conditions be `< 0`? `0` is reserved, but I think it's technically a "valid" port.

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+			shouldSwap = true
+		} else if cmp == 0 && srcPort > dstPort {
+			shouldSwap = true
+		}
+		if shouldSwap {
+			srcIPBytes, dstIPBytes = dstIPBytes, srcIPBytes
+			srcPort, dstPort = dstPort, srcPort
+		}
+
+		// Build the flow, format: <seed:2><protocol:1><src_ip><dst_ip><src_port:2><dst_port:2>
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]:
```suggestion
```

Let's move this comment to the function's docs.

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+		srcIPBytes: srcIPBytes,
+		dstIPBytes: dstIPBytes,
+		srcPort:    uint16(srcPort),
+		dstPort:    uint16(dstPort),
+		protocol:   protocolValue,
+		seed:       seedValue,
+	}, nil
+}
+
+func flow(srcIPBytes net.IP, srcPortForHash uint16, dstIPBytes net.IP, dstPortForHash uint16, protoValue uint8, seedValue uint16) string {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]:
```suggestion
// flow generates a flow tuple from the given parameters and returns it as a base64 encoded string
// format: <seed:2><src_ip><dst_ip><protocol:1><src_port:2><dst_port:2>
func flow(srcIPBytes net.IP, srcPortForHash uint16, dstIPBytes net.IP, dstPortForHash uint16, protoValue uint8, seedValue uint16) string {
```

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+		args.SourceIP,
+		args.SourcePort,
+		args.DestinationIP,
+		args.DestinationPort,
+		args.Protocol,
+		args.Seed,
+	), nil
+}
+
+type communityIDParams struct {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: I liked the struct idea, but I think we could better utilize it than just passing parameters around. For example, instead of `communityIDParams`, we could have `communityIDHash` and add functions to either swap the IPs and calculate the flow tuple hash, without having to pass all those parameters around. 

Considering all functions share the same package, it would be nice to make the struct act like that, so we don't pollute the package with generic function names like `flow` and `validateAndExtractParams`

WDYT?

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get source IP: %w", err)
+	}
+
+	dstIPValue, err := destinationIP.Get(ctx, tCtx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get destination IP: %w", err)
+	}
+
+	srcPort, err := sourcePort.Get(ctx, tCtx)
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]: The port and IP parsing/validation logic are repeating here, can we have a function for each that takes the port/ip argument and returns it's validated uint8/net.IP value?

File: pkg/ottl/ottlfuncs/func_community_id.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+var protocolMap = map[string]uint8{
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id.go]:
```suggestion
var communityIDProtocols = map[string]uint8{
```

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 7 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: `0` is mostly reserved, IANA also marks it as reserved as real world applications do not use it. So I am not seeing the value of allowing 0 here but open if we would like to make it general (maybe dev env can use, etc..)

Comment (edmocosta) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Yes, that's true. I'm comparing it to the reference implementation, which does allow zero ports ([see](https://github.com/corelight/pycommunityid/blob/master/communityid/algo.py#L271)), that's why I asked it. IMO, there's no issues allowing it here as well to comply with that implementation.

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 11 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Ah, initially tried to simplify into a method but naming was (still I don'n like `extractIPAndPort` kk) hard and just left it was... but made it with this commit that does validation and extraction - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/43398/commits/5037d545357d248257b8b3b0a0c997cad30e4b6e

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 10 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Yeah, initially didn't use struct and `validateAndExtractParams` became horrible where we cannot figure out what  values we are returning when we have error cases internally, like:
```
func validateAndExtractParams[K any](...) ([]byte, uint16, []byte, uint16, uint8, uint16, error) {
	srcIPAddr, srcPort, err := extractIPAndPort(ctx, tCtx, sourceIP, sourcePort, "source")
	if err != nil {
		return nil, 0, nil, 0, 0, 0, err // not good practice
	}
```
Then, as we have a standard struct for args (`CommunityIDArguments` for this case), introduced a struct.
To extend beyond, I agree with the point that instead of just being a bridge, adding meaningful operations would also make sense. So, it has (after renaming) now `normalize` (swaps if required) and `compute` (calculates flow hash) as the best practice.
Let me know if we can improve more.
Thanks.

File: pkg/ottl/ottlfuncs/func_community_id.go (hunk 7 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id.go]: Allowing zero doesn't hurt here as you mentioned and the upstream's point might be considering range of protocols (zero is mostly restricted for TCP & UDP AFAIK). So, allowed zero in our logic as well.

File: pkg/ottl/ottlfuncs/README.md [hunk 13]
```diff
@@ -574,6 +575,28 @@ Examples:
...
+
+`CommunityID(sourceIP, sourcePort, destinationIP, destinationPort, Optional[protocol], Optional[seed])`
+
+The `CommunityID` converter generates a network hash flow. Community ID is a standardized flow hashing algorithm that produces consistent hash values for network connections, useful when correlating network traffic across different monitoring systems.
+The output format is base64 encoding string of <2-byte-seed><source-IP-bytes><destination-IP-bytes><1-byte-protocol><2-byte-source-port><2-byte-destination-port>.
+
+`sourceIP` is a source IP address (IPv4 or IPv6).
+`sourcePort` is a source port (must be between 1 and 65535).
+`destinationIP` is a destination IP address (IPv4 or IPv6).
+`destinationPort` is a destination port (must be between 1 and 65535).
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/README.md]:
```suggestion
`destinationPort` is a destination port (must be between 0 and 65535).
```

File: pkg/ottl/ottlfuncs/README.md [hunk 14]
```diff
@@ -574,6 +575,28 @@ Examples:
...
 
+### CommunityID
+
+`CommunityID(sourceIP, sourcePort, destinationIP, destinationPort, Optional[protocol], Optional[seed])`
+
+The `CommunityID` converter generates a network hash flow. Community ID is a standardized flow hashing algorithm that produces consistent hash values for network connections, useful when correlating network traffic across different monitoring systems.
+The output format is base64 encoding string of <2-byte-seed><source-IP-bytes><destination-IP-bytes><1-byte-protocol><2-byte-source-port><2-byte-destination-port>.
+
+`sourceIP` is a source IP address (IPv4 or IPv6).
+`sourcePort` is a source port (must be between 1 and 65535).
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/README.md]:
```suggestion
`sourcePort` is a source port (must be between 0 and 65535).
```

File: pkg/ottl/ottlfuncs/func_community_id_test.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+			name:           "Invalid Protocol",
+			sourceIP:       "10.0.0.1",
+			destIP:         "10.0.0.2",
+			sourcePort:     80,
+			destPort:       12345,
+			protocol:       "UNKNOWN",
+			seed:           0,
+			errorExpected:  true,
+			expectedErrMsg: "unsupported protocol",
+		},
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_community_id_test.go:170]:
```suggestion
		},
		{
			name:          "Different IP families",
			sourceIP:      "2001:db8::1",
			destIP:        "10.1.1.1",
			sourcePort:    8080,
			destPort:      443,
			protocol:      "TCP",
			seed:          0,
			expected:      "1:T1e3va6jTSWaSKkG3k4dnZq5VM8=",
			errorExpected: false,
		},
		{
			name:          "Same IP lengths with higher source port",
			sourceIP:      "10.1.1.1",
			destIP:        "10.1.1.2",
			sourcePort:    8080,
			destPort:      8079,
			protocol:      "TCP",
			seed:          0,
			expected:      "1:1fjNfFLjLTsO2WE9AY0TbTeSaBs=",
			errorExpected: false,
		},
		{
			name:          "Same IP lengths with higher dest port",
			sourceIP:      "10.1.1.2",
			destIP:        "10.1.1.1",
			sourcePort:    8079,
			destPort:      8080,
			protocol:      "TCP",
			seed:          0,
			expected:      "1:1fjNfFLjLTsO2WE9AY0TbTeSaBs=",
			errorExpected: false,
		},
		{
			name:          "Same IP with higher source port",
			sourceIP:      "10.1.1.2",
			destIP:        "10.1.1.2",
			sourcePort:    8081,
			destPort:      8080,
			protocol:      "TCP",
			seed:          0,
			expected:      "1:3byUvmBpOu8KHoTRr4eN5AjTQEU=",
			errorExpected: false,
		},
		{
			name:          "Same IP with higher dest port",
			sourceIP:      "10.1.1.2",
			destIP:        "10.1.1.2",
			sourcePort:    8080,
			destPort:      8081,
			protocol:      "TCP",
			seed:          0,
			expected:      "1:3byUvmBpOu8KHoTRr4eN5AjTQEU=",
			errorExpected: false,
		},
```

Some extra edge-cases implemented but not covered by tests.

File: pkg/ottl/ottlfuncs/README.md (hunk 13 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/README.md]: Oops! Thanks for catching this up.

File: pkg/ottl/ottlfuncs/func_community_id_test.go (hunk 15 above)
Comment (mashhurs) [reply pkg/ottl/ottlfuncs/func_community_id_test.go:170]: Verified!
```
    --- PASS: TestCommunityID/Different_IP_families (0.00s)
    --- PASS: TestCommunityID/Same_IP_lengths_with_higher_source_port (0.00s)
    --- PASS: TestCommunityID/Same_IP_lengths_with_higher_dest_port (0.00s)
    --- PASS: TestCommunityID/Same_IP_with_higher_source_port (0.00s)
    --- PASS: TestCommunityID/Same_IP_with_higher_dest_port (0.00s)
```

**Comment by edmocosta** [comment]
Hi @mashhurs, could you please resolve the conflicts so we can move forward with this PR? Thanks!

**Comment by mashhurs** [comment]
> Hi @mashhurs, could you please resolve the conflicts so we can move forward with this PR? Thanks!

I have just did it but I am not sure about CI failure if it is related to my change - https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/19545598214/job/55963960572?pr=43398

**Review by edmocosta — APPROVED** [review]
